contracts/Tendermint-ORU.sol

```diff
@@ -0,0 +1,254 @@
+        _headerHeights[genesisSubmissionHash] = 1;
+
+        // Set the tip hash
+        _tipHash = genesisHash;
+    }
+
+    ////////////////////////////////////
+    // External functions
+    ////////////////////////////////////
+
+    /// @notice Submit a new bare block, placing a bond.
+    function submitBlock(BareBlock calldata bareBlock, bytes32 prevSubmissionHash) external payable {
+        // Must send _bondSize ETH to submit a block
+        require(msg.value == _bondSize);
+        // Previous block header hash must be the tip
+        require(bareBlock.header.lastBlockID == _tipHash);
+        // Height must increment
+        // Note: orphaned blocks be pruned before submitting new blocks since
+        // this check does not account for forks.
+        require(bareBlock.header.height == SafeMath.add(_headerHeights[prevSubmissionHash], 1));
+
+        // Take simple hash of commit for previous block
+        bytes32 lastCommitHash = keccak256(abi.encode(bareBlock.lastCommit));
+
+        // TODO serialize header
+        bytes memory serializedHeader;
+
+        // Hash serialized header
+        bytes32 headerHash = keccak256(serializedHeader);
+
+        // Insert header as new tip
+        HeaderSubmission memory headerSubmission = HeaderSubmission(
+            bareBlock.header,
+            msg.sender,
+            block.number,
+            lastCommitHash
+        );
+        bytes32 headerSubmissionHash = keccak256(abi.encode(headerSubmission));
+        _headerSubmissionHashes[headerHash] = headerSubmissionHash;
+        _headerHeights[headerSubmissionHash] = bareBlock.header.height;
+        _isNotFinalized[headerHash] = true;
+        _tipHash = headerHash;
+
+        emit BlockSubmitted(bareBlock, headerHash, bareBlock.header.height, headerSubmission);
+    }
+
+    /// @notice Prove a block was invalid, reverting it and orphaning its descendents.
+    function proveFraud(
+        bytes32 headerHash,
+        HeaderSubmission calldata headerSubmission,
+        HeaderSubmission calldata tipSubmission,
+        Commit calldata commit
+    ) external {
+        // Check submission against storage
+        bytes32 headerSubmissionHash = keccak256(abi.encode(headerSubmission));
+        require(headerSubmissionHash == _headerSubmissionHashes[headerHash]);
+        // Block must not be finalized yet
+        require(_isNotFinalized[headerHash]);
+
+        // Block must be at most the same height as the tip
+        // Note: orphaned blocks be pruned before submitting new blocks since
+        // this check does not account for forks.
+        require(keccak256(abi.encode(tipSubmission)) == _headerSubmissionHashes[_tipHash]);
+        require(headerSubmission.header.height <= tipSubmission.header.height);
+
+        // TODO serialize and Merkleize commit
+        // TODO compare root against stored root
+        // TODO process commit, check at least 2/3 of voting power
```

Comment:
Do you think this will be possible to do in one transaction or will it exceed the gas limit?

Reply:
Just by eyeballing, I would say "possible, but very tight." It depends on what serialization scheme is used, mostly. Each ecrecover is 3000 gas, so 100 of those would be a measly 300,000.